### DISCUSSION TITLE
Is it possible to disable the PEP 440 checks?

### DISCUSSION THREAD

**Discussion by drcongo** [python-poetry/poetry#6367] [Q&A] [answered]
I'm having a hell of a time trying to get poetry 1.2.0 running on Ubuntu, both on all our servers and in our Docker containers for CI. I'm currently trying to get it working in our base Docker image so I'll focus on the issues with that here for now. Having lost yesterday trying to install it with the new official installer on a server, I thought I'd try pipx in docker. I ended up having to add `/root/.local/pipx/venvs/poetry/bin/` to the path, which kinda makes sense, I was then able to run `pytest`, which was good, however the application itself was not runnable. Trying to start a Django dev server, or anything else really just errored out because the application couldn't see the packages that poetry had installed.

So, I returned to the official installer, added `/root/.local/bin/poetry` to the path and ran `poetry install` but got `Invalid PEP 440 version: '0.23ubuntu1'`, after digging around in the issues yesterday I'd discovered that this was some kind of problem with the system Python on Ubuntu, so I installed a new version from deadsnakes, rebuilt everything, ran poetry install and got `Invalid PEP 440 version: '0.23ubuntu1'` again. Given that this is a development container, there's no `venv` involved, the `0.23ubuntu1` is coming from a system package that I have no control over, and I'm not even invoking the system Python.

I've completely run out of ideas at this point, and I'd love to know what the official advice / position is. Can I turn off the PEP440 checks? 

My `apt list --installed | grep python` looks like this...

```
libpython2-stdlib/focal,now 2.7.17-2ubuntu4 arm64 [installed,automatic]
libpython2.7-minimal/focal-updates,focal-security,now 2.7.18-1~20.04.3 arm64 [installed,automatic]
libpython2.7-stdlib/focal-updates,focal-security,now 2.7.18-1~20.04.3 arm64 [installed,automatic]
libpython3-dev/focal,now 3.8.2-0ubuntu2 arm64 [installed,automatic]
libpython3-stdlib/focal,now 3.8.2-0ubuntu2 arm64 [installed,automatic]
libpython3.10-dev/focal,now 3.10.6-1+focal1 arm64 [installed,automatic]
libpython3.10-minimal/focal,now 3.10.6-1+focal1 arm64 [installed,automatic]
libpython3.10-stdlib/focal,now 3.10.6-1+focal1 arm64 [installed,automatic]
libpython3.10/focal,now 3.10.6-1+focal1 arm64 [installed,automatic]
libpython3.8-dev/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04.5 arm64 [installed,automatic]
libpython3.8-minimal/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04.5 arm64 [installed,automatic]
libpython3.8-stdlib/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04.5 arm64 [installed,automatic]
libpython3.8/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04.5 arm64 [installed,automatic]
python-apt-common/focal-updates,now 2.0.0ubuntu0.20.04.8 all [installed,automatic]
python-cffi-backend/focal,now 1.14.0-1build1 arm64 [installed,automatic]
python-cryptography/focal-updates,focal-security,now 2.8-3ubuntu0.1 arm64 [installed,automatic]
python-enum34/focal,now 1.1.6-2ubuntu1 all [installed,automatic]
python-ipaddress/focal,now 1.0.17-1build1 all [installed,automatic]
python-openssl/focal,now 19.0.0-1build1 all [installed]
python-pip-whl/focal-updates,now 20.0.2-5ubuntu1.6 all [installed,automatic]
python-six/focal,now 1.14.0-2 all [installed,automatic]
python-sqlite/focal,now 1.0.1-12ubuntu2 arm64 [installed]
python2-minimal/focal,now 2.7.17-2ubuntu4 arm64 [installed,automatic]
python2.7-minimal/focal-updates,focal-security,now 2.7.18-1~20.04.3 arm64 [installed,automatic]
python2.7/focal-updates,focal-security,now 2.7.18-1~20.04.3 arm64 [installed,automatic]
python2/focal,now 2.7.17-2ubuntu4 arm64 [installed,automatic]
python3-apt/focal-updates,now 2.0.0ubuntu0.20.04.8 arm64 [installed,automatic]
python3-certifi/focal,now 2019.11.28-1 all [installed,automatic]
python3-chardet/focal,now 3.0.4-4build1 all [installed,automatic]
python3-dbus/focal,now 1.2.16-1build1 arm64 [installed,automatic]
python3-dev/focal,now 3.8.2-0ubuntu2 arm64 [installed]
python3-distro-info/focal,now 0.23ubuntu1 all [installed,automatic]
python3-distro/focal,now 1.4.0-1 all [installed,automatic]
python3-distutils/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04 all [installed]
python3-gi/focal,now 3.36.0-1 arm64 [installed,automatic]
python3-idna/focal,now 2.8-1 all [installed,automatic]
python3-lib2to3/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04 all [installed,automatic]
python3-libsass/focal,now 0.18.0-1build1 arm64 [installed]
python3-minimal/focal,now 3.8.2-0ubuntu2 arm64 [installed,automatic]
python3-mysqldb/focal,now 1.4.4-2build1 arm64 [installed]
python3-olefile/focal,now 0.46-2 all [installed,automatic]
python3-pil/focal-updates,focal-security,now 7.0.0-4ubuntu0.5 arm64 [installed]
python3-pip/focal-updates,now 20.0.2-5ubuntu1.6 all [installed]
python3-pkg-resources/focal,now 45.2.0-1 all [installed,automatic]
python3-psycopg2/focal,now 2.8.4-2 arm64 [installed]
python3-pycurl/focal,now 7.43.0.2-1ubuntu5 arm64 [installed]
python3-pygments/focal-updates,focal-security,now 2.3.1+dfsg-1ubuntu2.2 all [installed,automatic]
python3-requests-unixsocket/focal,now 0.2.0-2 all [installed,automatic]
python3-requests/focal,now 2.22.0-2ubuntu1 all [installed,automatic]
python3-setuptools/focal,now 45.2.0-1 all [installed]
python3-six/focal,now 1.14.0-2 all [installed,automatic]
python3-software-properties/focal-updates,now 0.99.9.8 all [installed,automatic]
python3-urllib3/focal-updates,focal-security,now 1.25.8-2ubuntu0.1 all [installed,automatic]
python3-wheel/focal,now 0.34.2-1 all [installed,automatic]
python3-yaml/focal-updates,focal-security,now 5.3.1-1ubuntu0.1 arm64 [installed,automatic]
python3.10-dev/focal,now 3.10.6-1+focal1 arm64 [installed]
python3.10-distutils/focal,now 3.10.6-1+focal1 all [installed]
python3.10-lib2to3/focal,now 3.10.6-1+focal1 all [installed,automatic]
python3.10-minimal/focal,now 3.10.6-1+focal1 arm64 [installed,automatic]
python3.10-venv/focal,now 3.10.6-1+focal1 arm64 [installed]
python3.10/focal,now 3.10.6-1+focal1 arm64 [installed]
python3.8-dev/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04.5 arm64 [installed]
python3.8-minimal/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04.5 arm64 [installed,automatic]
python3.8-venv/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04.5 arm64 [installed]
python3.8/focal-updates,focal-security,now 3.8.10-0ubuntu1~20.04.5 arm64 [installed]
python3/focal,now 3.8.2-0ubuntu2 arm64 [installed]
```

Note that the `0.23ubuntu1` version comes from `python3-distro-info` which has been automatically installed. There's a good chance that I'm completely misunderstanding how we're supposed to use Poetry 1.2.0, and I'm just trying to carry on using it the way that has always worked, especially given that there's no discussions about this already. However, I can't find any information at all about any of this. The closest the announcement blog post comes is the line 

> Standard install locations are used for both Poetry itself, and the poetry CLI wrapper.

which doesn't mention what those locations are or what's standard about them (I've always thought `/usr/local/bin/` where it used to live was standard).

Any help with this would be hugely appreciated, I've spent two days on it now and gotten nowhere and we're really going to need the CI back up soon. Thanks.

**Comment by neersighted** [comment]
Hi @drcongo -- the solution here is to use a virtual environment. Poetry needs to be able to reason about and manage the environment your project is installed into. One of the reasons for this is what you've run into -- distro packaging may result in packages that are invalid according to the rules of the Python ecosystem that Poetry adheres to (and Poetry needs to understand what packages and versions are in your environment in order to install/upgrade/uninstall to meet your project's needs). Similarly, there may be conflicts between your project's dep tree and the Python deps present in the system (say you `apt install <python-based tool>` and then try to have Poetry install to the system environment).

Virtual environments are a powerful tool and an expected part of using Python, even in a container. I'm sorry that what was working before no longer works -- however it was only ever a happy coincidence that it did, and not an intended behavior. We're working on better documentation/resources for this (there was a user who wanted to spearhead documentation and guidance for containers, but they seem to have disappeared) -- if you would like to participate you would certainly be welcome.

**Comment by drcongo** [comment]
OK, thanks @neersighted, I'll try adding a `venv` in the containers. However, I did get the exact same PEP440 error yesterday on a server using a venv so the question of whether we can disable the check still stands.

**Reply by neersighted** [reply]
A minimal reproduction would be great -- as it stands right now I can install and use Poetry just fine in a Ubuntu 22.04 container with Python 3.10 as best I can tell (complications with 3.10 packaging and `DEB_PYTHON_INSTALL_LAYOUT` aside -- 3.9 works with no complications).

Also for what it's worth, it's not really that we're 'checking' or 'enforcing' PEP 440 -- it's that Poetry 1.2 is a lot smarter about upgrades and managing the environment, and that requires the regular version parsing code. The best we could do is ignore packages that fail to parse, but that would most likely lead to a lot of unexpected or hard to troubleshoot circumstances. Since this really only comes up when trying to install to the system environment (which is fraught), it seems like a footgun easily avoided.

**Reply by drcongo** [reply]
Ugh, yeah, that does sound painful.

**Comment by drcongo** [comment]
Update: This is what happens if I use a venv in the container...
```
~ poetry install
Creating virtualenv *** in /usr/srv/app/.venv
Installing dependencies from lock file

Package operations: 111 installs, 0 updates, 0 removals

[...]
  • Installing pytest (6.2.5)
[...]

~ pytest
Unknown command: pytest
~ which pytest
 
~ poetry show pytest
 name         : pytest
 version      : 6.2.5
 description  : pytest: simple powerful testing with Python

[...]
```

**Reply by drcongo** [reply]
OK, I've stripped out a whole bunch of stuff from our Dockerfile and the above no longer happens, so I'm getting somewhere. I've no idea which bit of it was causing the problem though. If I work it out I'll report back.

**Comment by drcongo** [comment]
So close, but now I have one of the problems I had yesterday in that poetry creates a `venv` with no `bin` directory so it's impossible to source it. Is this a new feature too?

"Minimal" Docker file here: 

```
FROM ubuntu:20.04 as base

ENV VERSION 2.0.0

# set noninteractive installation
RUN mkdir -p /etc/ssl/certs/java/ && \
    export DEBIAN_FRONTEND=noninteractive && \
    apt-get update && \
    apt-get upgrade -y && \
    apt-get install -y \
    tzdata \
    language-pack-en && \
    ln -fs /usr/share/zoneinfo/Europe/London /etc/localtime && \
    dpkg-reconfigure --frontend noninteractive tzdata && \
    update-locale LANG=en_GB.UTF-8 LC_MESSAGES=POSIX


RUN apt-get update && \
    apt-get upgrade -y && \
    apt-get install -y \
    autojump \
    automake \
    bash \
    build-essential \
    bzip2 \
    ca-certificates \
    checkinstall \
    curl \
    default-jre \
    fish \
    g++ \
    gcc \
    gettext \
    git \
    libbz2-dev \
    libc-dev \
    libc6-dev \
    libcairo2 \
    libcairo2-dev \
    libcurl4 \
    libcurl4-openssl-dev \
    libevent-2.1-7 \
    libevent-dev \
    libffi-dev \
    libfreetype6 \
    libfreetype6-dev \
    libgdbm-dev \
    libgdk-pixbuf2.0-0 \
    libjpeg-dev \
    libjpeg8 \
    libjpeg8-dev \
    liblzma-dev \
    libncurses5-dev \
    libncursesw5-dev \
    libpango1.0-0 \
    libpcre3-dev \
    libpng-dev \
    libpq-dev \
    libreadline-dev \
    libsass-dev \
    libsass1 \
    libsqlite3-dev \
    libssl-dev \
    libxml2-dev \
    libxslt1-dev \
    llvm \
    make \
    nano \
    nodejs \
    openssh-server \
    postgresql-12 \
    python-openssl \
    python-sqlite \
    python3-dev \
    python3-libsass \
    python3-mysqldb \
    python3-pil \
    python3-psycopg2 \
    python3-pycurl \
    python3-setuptools \
    python3.8-venv \
    rsync \
    software-properties-common \
    sqlite \
    sudo \
    tk-dev \
    wget \
    xz-utils \
    zlib1g-dev \
    zsh \
    && rm -rf /root/.cache && \
    rm -rf /var/lib/apt/lists/*

ENV SHELL /bin/zsh


FROM base AS app

# Install and set up Poetry for python dependencies management
# Apparently we now need to have a .venv in the container
ENV POETRY_VIRTUALENVS_CREATE=1 \
    POETRY_VIRTUALENVS_IN_PROJECT=1 \
    POETRY_VIRTUALENVS_PATH="/usr/venvs" \
    POETRY_REPOSITORIES={}

# Install poetry
RUN curl -sSL https://install.python-poetry.org | python3 - && \
    ln -s /usr/bin/python3 /usr/bin/python && \
    ln -s /root/.local/bin/poetry /usr/local/bin/poetry && \
    python -m venv .venv && \
    mkdir -p /usr/srv/app/


WORKDIR /usr/srv/app/

CMD "sleep 99999999"
```

This is the base image that a lot of our application images build on.

**Comment by neersighted** [comment, marked as answer]
I'm not sure what `python -m venv .venv` is supposed to accomplish with its present placement -- anyway, try the following:

```dockerfile
FROM ubuntu:20.04 as base

# set noninteractive installation
RUN mkdir -p /etc/ssl/certs/java/ && \
    export DEBIAN_FRONTEND=noninteractive && \
    apt-get update && \
    apt-get upgrade -y && \
    apt-get install -y \
    tzdata \
    language-pack-en && \
    ln -fs /usr/share/zoneinfo/Europe/London /etc/localtime && \
    dpkg-reconfigure --frontend noninteractive tzdata && \
    update-locale LANG=en_GB.UTF-8 LC_MESSAGES=POSIX


RUN apt-get update && \
    apt-get upgrade -y && \
    apt-get install -y \
    autojump \
    automake \
    bash \
    build-essential \
    bzip2 \
    ca-certificates \
    checkinstall \
    curl \
    default-jre \
    fish \
    g++ \
    gcc \
    gettext \
    git \
    libbz2-dev \
    libc-dev \
    libc6-dev \
    libcairo2 \
    libcairo2-dev \
    libcurl4 \
    libcurl4-openssl-dev \
    libevent-2.1-7 \
    libevent-dev \
    libffi-dev \
    libfreetype6 \
    libfreetype6-dev \
    libgdbm-dev \
    libgdk-pixbuf2.0-0 \
    libjpeg-dev \
    libjpeg8 \
    libjpeg8-dev \
    liblzma-dev \
    libncurses5-dev \
    libncursesw5-dev \
    libpango1.0-0 \
    libpcre3-dev \
    libpng-dev \
    libpq-dev \
    libreadline-dev \
    libsass-dev \
    libsass1 \
    libsqlite3-dev \
    libssl-dev \
    libxml2-dev \
    libxslt1-dev \
    llvm \
    make \
    nano \
    nodejs \
    openssh-server \
    postgresql-12 \
    python-openssl \
    python-sqlite \
    python3-dev \
    python3-libsass \
    python3-mysqldb \
    python3-pil \
    python3-psycopg2 \
    python3-pycurl \
    python3-setuptools \
    python3.8-venv \
    rsync \
    software-properties-common \
    sqlite \
    sudo \
    tk-dev \
    wget \
    xz-utils \
    zlib1g-dev \
    zsh \
    && rm -rf /root/.cache && \
    rm -rf /var/lib/apt/lists/*

ENV SHELL /bin/zsh

FROM base AS app

ENV POETRY_HOME="/opt/poetry" \
    POETRY_VIRTUALENVS_IN_PROJECT=1

# Install poetry
RUN curl -sSL https://install.python-poetry.org | python3 - --version 1.2.0 && \
    ln -s $POETRY_HOME/bin/poetry /usr/local/bin/poetry

# Other setup
RUN ln -s /usr/bin/python3 /usr/bin/python && \
    mkdir -p /usr/srv/app/

WORKDIR /usr/srv/app/

CMD "sleep 99999999"
```

Note some key differences:
* `POETRY_HOME` can be used with the official installer to set the install path to somewhere system-wide.
* Pinned versions are best practice in CI.
* There's no need to tell Poetry to create virtual environments -- we do so by default. I think the `python -m venv .venv` was a misguided attempt to create it manually? If you're going to use Poetry's built-in env management, you're better off fully using it.
* I'm not sure why you set `POETRY_REPOSITORIES={}`, so drop that.
* Likewise, `POETRY_VIRTUALENVS_PATH` doesn't do anything useful when you set `POETRY_VIRTUALENVS_IN_PROJECT`.

For a clean start, I would suggest something like this:

```dockerfile
ENV POETRY_HOME="/opt/poetry"

RUN python3 -m venv $POETRY_HOME && \
    $POETRY_HOME/bin/pip install poetry==1.2.0

RUN git clone <repo> /usr/srv/app

WORKDIR /usr/srv/app

ENV VIRTUAL_ENV=/usr/srv/app-env \
    PATH="/usr/srv/app-env/bin:${POETRY_HOME}/bin:${PATH}"

RUN python3 -m venv $VIRTUAL_ENV && \
    poetry install
    
CMD "python -m <myapp>" # or
CMD "<my script entrypoint>" # etc
```

**Reply by drcongo** [reply]
Thanks again, hugely appreciated. The `POETRY_VIRTUALENVS_PATH` etc. where just still in there from the old version. The `python -m venv .venv` was a poorly thought through attempt to have it create a `venv` with a `bin` directory in it. I've since moved that to the application Dockerfiles immediately before running `poetry install` and fingers crossed, it seems to work. The pinning we do in the individual application Dockerfiles for various reasons with `poetry self update 1.2.0` etc. The `POETRY_HOME` tip is a really useful one. 

Thanks again. Hopefully I can leave you alone now!

**Reply by neersighted** [reply]
Check this out: this is my minimal best practices way to run the Poetry test suite:

```dockerfile
FROM python:3.10

ENV POETRY_HOME=/opt/poetry
RUN python3 -m venv $POETRY_HOME && \
    $POETRY_HOME/bin/pip install poetry==1.2.0

RUN git clone https://github.com/python-poetry/poetry /poetry

WORKDIR /poetry

ENV VIRTUAL_ENV=/poetry-env \
    PATH="/poetry-env/bin:$POETRY_HOME/bin:$PATH"

RUN python3 -m venv $VIRTUAL_ENV && \
    poetry install
    
CMD "pytest"
```

Hopefully this real example of my 'clean' pattern above is useful (also see the [docs](https://python-poetry.org/docs/#ci-recommendations)).
